networkx/classes/tests/dispatch_interface.py

```diff
@@ -108,71 +110,92 @@ def convert_from_nx(
-                        u,
-                        v,
-                        key,
-                        {
-                            k: datadict.get(k, default)
-                            for k, default in edge_attrs.items()
-                            if default is not None or k in datadict
-                        },
-                    )
-                    for u, nbrs in graph._adj.items()
-                    for v, keydict in nbrs.items()
-                    for key, datadict in keydict.items()
-                )
-            else:
-                G.add_edges_from(
-                    (u, v, key, {})
-                    for u, nbrs in graph._adj.items()
-                    for v, keydict in nbrs.items()
-                    for key, datadict in keydict.items()
-                )
-        elif preserve_edge_attrs:
-            G.add_edges_from(graph.edges(data=True))
+
+        # tools to build datadict and keydict
+        if preserve_edge_attrs:
+
+            def G_new_datadict(old_dd):
+                return G.edge_attr_dict_factory(old_dd)
         elif edge_attrs:
-            G.add_edges_from(
-                (
-                    u,
-                    v,
-                    {
-                        k: datadict.get(k, default)
-                        for k, default in edge_attrs.items()
-                        if default is not None or k in datadict
-                    },
+
+            def G_new_datadict(old_dd):
+                return G.edge_attr_dict_factory(
+                    (attr, old_dd.get(attr, default))
+                    for attr, default in edge_attrs.items()
+                    if default is not None or attr in old_dd
                 )
-                for u, v, datadict in graph.edges(data=True)
-            )
         else:
-            G.add_edges_from(graph.edges)
+
+            def G_new_datadict(old_dd):
+                return G.edge_attr_dict_factory()
+
+        if G.is_multigraph():
+
+            def G_new_d(keydict):
+                kd = G.adjlist_inner_dict_factory(
+                    (k, G_new_datadict(dd)) for k, dd in keydict.items()
+                )
+                return kd
+        else:
+            G_new_d = G_new_datadict
+
+        # add edges
+        G_adj = G._adj
+        if G.is_directed():
+            for n, nbrs in graph._adj.items():
+                G_adj[n].update((nbr, G_new_d(dd)) for nbr, dd in nbrs.items())
+            # ensure same datadict for pred and adj; and pred order of graph._pred
+            for n, nbrs in graph._pred.items():
+                G._pred[n].update((nbr, G_adj[nbr][n]) for nbr in nbrs)
+        else:  # undirected
+            for n, nbrs in graph._adj.items():
+                # ensure same datadict for both ways; and adj order of graph._adj
+                G_adj[n].update(
+                    (nbr, G_adj[nbr].get(n, G_new_d(dd))) for nbr, dd in nbrs.items()
+                )
+
+        #        # add edges
```

Comment:
I'm not as familiar with NetworkX conventions as others, but is there a reason this much code would be commented out and not just removed?